File: Solution MinHeap.cpp

import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

public class NumberContainers {

    private final Map<Integer, Integer> indexToValue;
    private final Map<Integer, PriorityQueue<Integer>> valueToIndex;
    private static final int NOT_FOUND = -1;

    public NumberContainers() {
        indexToValue = new HashMap<>();
        valueToIndex = new HashMap<>();
    }

    public void change(int index, int number) {
        indexToValue.put(index, number);
        valueToIndex.putIfAbsent(number, new PriorityQueue<>());
        valueToIndex.get(number).add(index);
    }

    public int find(int number) {
        if (!valueToIndex.containsKey(number)) {
            return NOT_FOUND;
        }

        PriorityQueue<Integer> indexesForCurrentValue = valueToIndex.get(number);
        while (!indexesForCurrentValue.isEmpty()) {
            int index = indexesForCurrentValue.peek();
            if (indexToValue.get(index) == number) {
                return index;
            }
            indexesForCurrentValue.poll();
        }
        return NOT_FOUND;
    }
}

File: Solution MinHeap.js

var NumberContainers = function () {
    const {PriorityQueue} = require('@datastructures-js/priority-queue');
    this.indexToValue = new Map();//Map <Number, Number>
    this.valueToIndex = new Map();//Map <Number, MinPriorityQueue<Number>>
    this.NOT_FOUND = -1;
};

/** 
 * @param {number} index 
 * @param {number} number
 * @return {void}
 */
NumberContainers.prototype.change = function (index, number) {
    this.indexToValue.set(index, number);
    if (!this.valueToIndex.has(number)) {
        this.valueToIndex.set(number, new MinPriorityQueue({compare: (x, y) => x - y}));
    }
    this.valueToIndex.get(number).enqueue(index);
};

/** 
 * @param {number} number
 * @return {number}
 */
NumberContainers.prototype.find = function (number) {
    if (!this.valueToIndex.has(number)) {
        return this.NOT_FOUND;
    }

    const indexesForCurrentValue = this.valueToIndex.get(number);
    while (!indexesForCurrentValue.isEmpty()) {
        const index = indexesForCurrentValue.front();
        if (this.indexToValue.get(index) === number) {
            return index;
        }
        indexesForCurrentValue.dequeue();
    }
    return this.NOT_FOUND;
};
